Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,37 +1,345 @@
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <stdbool.h>
 
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
+// Defined limit for a string key length 
+#define LIMIT 10
 
-// Hash struct
-typedef struct Hash {
-    int key;
+// Hash Item struct
+typedef struct Item {
+    char* key;
     int value;
-} Hash;
+} Item;
 
-void print() {
+// Hash Table struct
+typedef struct HashTable {
+    int size;
+    Item** items;
+} HashTable;
 
+
+// returns true if a collision would happen for the initially hashed index for a key/value pair and false if no collision would occurred 
+bool isCollision(HashTable * table, int index) {
+    if (table->items[index] != NULL) {
+        return true;
+    }
+    return false;
 }
 
-void traverse() {
+// equals checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool equals(HashTable * testHash, HashTable * expectedHash) {
+    for (int i = 0; i < expectedHash->size; i++) {
+        if (strcmp(testHash->items[i]->key, expectedHash->items[i]->key) != 0 || testHash->items[i]->value != expectedHash->items[i]->value) {
+            return false;
+        }
+    }
+    return true;
+}
 
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
+// Explanation:  I used loselose because it was a simple hashing function that seemed adequate for the purposes of this exercise.
+int hashing(HashTable *hashtable, char *key ) {
+    int i = 0;
+    int count = 0;
+    while (key[i] != NULL) {
+        count += key[i];
+        i++;
+    }
+    return count % hashtable->size;
 }
 
-bool compare() {
-    return true;
-    return false;
+
+// creates a new Item struct
+// returns new struct
+Item * createItem(char key[], int value, int * success) {
+    Item * item = malloc(sizeof(Item));
+    if (item != NULL) {        
+        item->key = strdup(key);
+        item->value = value;
+        *success = SUCCESS;
+        return item;
+    }
+    return NULL;
 }
 
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
+    HashTable * hashTable = malloc(sizeof(HashTable));
+    hashTable->items = malloc(sizeof(Item) * size);
```

Comment:
I think if the previous malloc failed, this one will crash before the check below will have a chance to run

Reply:
Agreed - added a line to catch after hashTable malloc. 
[commit c3a048af62cac3ffbed9b8b303e4041363916459](https://github.com/kcarrel/Training-Notes/pull/12/commits/c3a048af62cac3ffbed9b8b303e4041363916459)